wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,279 @@
+                                maxInputs
+                                (toTxOuts payees)
+                                availableUtxo
+    case res of
+         Left err -> return . Left . CoinSelectionFailed $ err
+         Right t  -> return . Right $ t
+
+-- | Special monad used to process the payments, which randomness is derived
+-- from a fixed seed obtained from hashing the payees. This guarantees that
+-- when we estimate the fees and later create a transaction, the coin selection
+-- will always yield the same value, making the process externally-predicatable.
+
+newtype PayMonad a = PayMonad {
+      buildPayment :: ReaderT (NonEmpty (Address, Coin)) IO a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadReader (NonEmpty (Address, Coin)))
+
+-- | \"Invalid\" 'MonadRandom' instance for 'PayMonad' which generates
+-- randomness using the hash of 'NonEmpty (Address, Coin)' as fixed seed.
+instance MonadRandom PayMonad where
+    getRandomBytes _ =
+        ask >>= return . ByteArray.convert
+                       . encodeUtf8 @Text @ByteString
+                       . sformat build
+                       . hash
+
+{-------------------------------------------------------------------------------
+  Estimating fees
+-------------------------------------------------------------------------------}
+
+data EstimateFeesError = EstFeesTxCreationFailed NewTransactionError
+
+instance Buildable EstimateFeesError where
+    build (EstFeesTxCreationFailed newTxErr) =
+        bprint ("EstFeesTxCreationFailed " % build) newTxErr
+
+instance Arbitrary EstimateFeesError where
+    arbitrary = EstFeesTxCreationFailed <$> arbitrary
+
+estimateFees :: ActiveWallet
+             -> IO Address
+             -- ^ A computation to generate a new change 'Address.
+             -> (Address -> Either CoinSelHardErr SafeSigner)
+             -- ^ A function to sign each Address
+             -> CoinSelectionOptions
+             -- ^ The options describing how to tune the coin selection.
+             -> HdAccountId
+             -- ^ The source @(root, account) from where the payment was
+             -- originated
+             -> NonEmpty (Address, Coin)
+             -- ^ The payees
+             -> IO (Either EstimateFeesError Coin)
+estimateFees activeWallet@ActiveWallet{..} genChangeAddr signAddress options accountId payees = do
+    snapshot         <- getWalletSnapshot walletPassive
+    let originalUtxo = accountUtxo snapshot accountId
+    res <- newTransaction activeWallet genChangeAddr signAddress options accountId payees
+    case res of
+         Left e  -> return . Left . EstFeesTxCreationFailed $ e
+         Right tx -> -- calculate the fee as the difference between inputs and outputs.
+             -- NOTE(adn) Apparently we shouldn't worry about the 'ExpenseRegulation'
+             -- affecting the way we sum, as no matter who pays for the fee,
+             -- the final difference won't be affected.
```

Comment:
Hmmm I don't think that's _entirely_ true. For sender pays fees we may need to add additional inputs (where recipient pays fees would just fail). But it's close enough I suppose.

Reply:
Agreed, this needs a better comment to explain things properly.